src/mimir_config.py

```diff
@@ -84,11 +87,11 @@ class Alertmanager(BaseModel):
 
 
 class _S3StorageBackend(BaseModel):
-    endpoint: str
-    access_key_id: str
-    secret_access_key: str
-    insecure: bool = False
-    signature_version: str = "v4"
+    access_key: str = ""
```

Comment:
if these are mandatory fields, as I expect, they shouldn't have default values. Am I missing something?

Reply:
The default values are used to facilitate checking if the model instance has been initialized with actual values. This allows us to determine if the S3 configuration is connected or not.

```
    def _get_s3_storage_config(self):
        """Retrieve S3 storage configuration."""
        if not self.s3_requirer.relations:
            return _S3ConfigData()
        raw = self.s3_requirer.get_s3_connection_info()
        raw["access_key"], raw["secret_key"] = raw.pop("access-key", ""), raw.pop("secret-key", "")
        return _S3ConfigData(**raw)
```

the above is triggered whenever a new worker joins the cluster, returning the model with default values helps in telling that Mimir isn't using S3 hence blocking the state

Reply:
but still, are we sure that the s3 databag is allowed to not contain any of these fields? If the databag is empty, it's likely not valid and we should abort.
Otherwise we'll call build_config with empty strings all over and we'll generate an invalid config.
I suggest you use an approach like [this](https://github.com/canonical/tempo-k8s-operator/blob/main/lib/charms/tempo_k8s/v1/tracing.py#L125) to validate the databags. Usually we do this in the charm lib but in this case it's an external library... Perhaps worth adding a separate `s3_interface.py` module to contain this?